### DISCUSSION TITLE
CodeIgniter Shield RESTful API

### DISCUSSION THREAD

**Discussion by ghost** [codeigniter4/shield#697] [Ideas]
Shield currently provides two methods of authentication, session and access tokens. There is already boilerplate controllers for authenticating via session. I think there could be interest in having a controller available "out of box" to provide a proper RESTful API for access token authentication.

**Comment by lonnieezell** [comment]
What type of methods would you like to see in that controller? 

Most of the auth token authentication is handled within the `tokens` filter.

**Reply by ghost** [reply]
I was thinking of an API that supports the functionality that the current view controllers have and restructure the code so that it is reusable for both the API and the view controllers (e.g. validation rules).

I believe a base for API authentication would be great. Right now there is controllers for registering and logging in from a form and can be used to protect content. This should be available in symmetry for an API.

I am working on just this right now for a personal project

Any suggestions into the implementation?

Also, if this idea is accepted into Shield I can make contributions to it.

**Reply by lonnieezell** [reply]
I'm confused. Do you mean login/registration using the Access Tokens? If so, the problem I can see there is that there are several ways that auth tokens could be handed out, and most don't work well in a traditional register/login flow. 

Or do you mean supporting username/email/whatever/password login/registration that we currently support handled via an API that we then use in our existing forms? I'm more in favor of that but then you start wading into the Javascript wars when deciding how to implement the views, though that could be done in plain javascript. I also had the thought of converting those forms to View Cells so they're more easily re-usable and able to be part of a dropdown or modal, which could potentially benefit from this design pattern or be in conflict with it, depending on how it all goes.

**Reply by sammyskills** [reply]
I think I understand what @miguel-rn means. Just like we have the `LoginController`, `RegisterController` and `MagicLinkController` for the web views, we should also have them for API authentication, maybe in a subfolder called **Api**.

With this, questions like:
- how to log a user in?
- how to register a new user?
- how to logout?
- how to activate a user _(if enabled)_?
- how to check if a user is authenticated?
- what settings in Shield are required for API authentication ONLY?
- `auth()` and `auth('tokens')`, what are the differences?

will be answered, making it easier to use and/or extend. Also, the documentation for mobile_apps and access tokens will need to be updated, as it seems that currently, shield focuses more on web authentication.

**Reply by lonnieezell** [reply]
Access tokens don't really have a login/logout scenario. They're intended for stateless usage, and the raw token is only ever made available when it's generated so either:

1. it's a semi-automated process like the [Mobile App Guide](https://github.com/codeigniter4/shield/blob/develop/docs/guides/mobile_apps.md) goes over, or
2. it's handled within the app itself where a user can manage their Access Tokens and will have to copy the raw token to a safe place when it is generated. 

There is also no activation flow as that would be handled with a traditional registration to your site/API which would handle activation. 

The docs already describe how to check for authenticated user (with the `auth()->user()` or `user()` helpers. 

The other items should be described in the app or the guides already I believe. If there's a hole in the docs we can fix it there. 

I believe the issue here is a misunderstanding of how Access Tokens are used, which we could maybe do a better job of clarifying. 

In short - I don't think there is a one-size-fits-all solution for Access Tokens, which is why a controller doesn't exist currently. 

It sounds like what's being expected is a flow more akin to what a JWT would provide, which we don't currently provide support for out of the box.

**Reply by ghost** [reply]
To clarify, I was thinking of controller(s) with API endpoints that can:

 - generate access tokens via credentials
 - revoke access tokens (I have not checked - is this currently supported? Or expiry?)
 - register user
 - updating user data (could accept optional user ID where users with permissions can update other users, else other users are only allowed to update themselves.)
 - changing of roles (again can implement role permission for this)

Although in a project I would favour Javascript for forms, I think we should keep the controllers as-is for simplicity (boilerplate example controllers). Except validation rules should be saved so they can be reused by both the view controllers and API.

**Reply by sammyskills** [reply]
> generate access tokens via credentials

Can you explain what you mean by credentials? I ask because there's already a way to generate access tokens by supplying the `device_name`.

> revoke access tokens (I have not checked - is this currently supported? Or expiry?)

Yes, this is already supported. `$user->revokeAccessToken($rawToken)`. See [docs](https://github.com/codeigniter4/shield/blob/develop/docs/guides/api_tokens.md#revoking-tokens).

> register user

I think this should follow similar implementation like the one for `RegisterController`. https://github.com/codeigniter4/shield/blob/52126e4c8c6f818db1ef92b8e7cc2b908973cade/src/Controllers/RegisterController.php#L85

> updating user data (could accept optional user ID where users with permissions can update other users, else other users are only allowed to update themselves.)

See [docs](https://github.com/codeigniter4/shield/blob/develop/docs/quickstart.md#editing-a-user).

> changing of roles (again can implement role permission for this)

See [docs](https://github.com/codeigniter4/shield/blob/develop/docs/authorization.md#managing-user-groups).

From what I understand, every other thing follows the same process apart from the handling of user authentication via access tokens.

**Reply by sammyskills** [reply]
> Access tokens don't really have a login/logout scenario. They're intended for stateless usage, and the raw token is only ever made available when it's generated so either...

If I am creating an API for a mobile app, at least, a logout endpoint should be provided. This I have seen with a lot of apps. 
What I expect this endpoint to do, is to:
1. Delete the current "session" _(if this is correct)_,
2. Revoke the access token, so the user will need to generate a new one for the next login.

> I believe the issue here is a misunderstanding of how Access Tokens are used, which we could maybe do a better job of clarifying.

Yes, I totally agree with you. With this, devs will be able to determine if shield fits the scope of what they want to build or not.

I am currently building an API for the mobile app version of a website built using shield, so, I'm wondering, based on the endpoints that I want to create, will shield do this for me, or should I just use [PHP JSON Web Tokens](https://github.com/firebase/php-jwt) ?

**Reply by ghost** [reply]
> > generate access tokens via credentials
> 
> Can you explain what you mean by credentials? I ask because there's already a way to generate access tokens by supplying the `device_name`.
> 

This would authenticate a user and if valid, return an access token. This would be used for a mobile app to login.


> > Access tokens don't really have a login/logout scenario. They're intended for stateless usage, and the raw token is only ever made available when it's generated so either...
> 
> If I am creating an API for a mobile app, at least, a logout endpoint should be provided. This I have seen with a lot of apps. 
> What I expect this endpoint to do, is to:
> 1. Delete the current "session" _(if this is correct)_,
> 2. Revoke the access token, so the user will need to generate a new one for the next login.
> 

To clarify, there is no session used with access tokens. And yes, this is why there should endpoints to provide this functionality for access tokens.

**Reply by lonnieezell** [reply]
I can see this working if we included those routes under a separate function. So, instead of `auth()->routes()`, we would have `auth()->apiRoutes()`, and then be able to specify the type of routes, like `mobile` or something. That makes it easier for people to opt-in to routes and not leave it wide open unintentionally.

**Reply by sammyskills** [reply]
> This would authenticate a user and if valid, return an access token. This would be used for a mobile app to login.

An example has already been given about this. See [Issuing the Tokens](https://github.com/codeigniter4/shield/blob/develop/docs/guides/mobile_apps.md#issuing-the-tokens).

**Reply by sammyskills** [reply]
Hi @lonnieezell,

Is this something already in the pipeline or your Todo list? If not, I'll suggest you help out with answers to the questions below, so someone else (@miguel-rn or @sammyskills) can take it up.
- What are the required Shield setup instructions for API implementation?
- Is the registration/activation process different from the site/web/traditional flow?
- When/where is the appropriate time/place to use `auth('tokens')` instead of `auth()` ?
- What is/are the appropriate way(s) to log a user out of the application completely?

@miguel-rn, if there are other questions that needs to be included, please do so.

Thank you.

**Comment by lonnieezell** [comment]
@sammyskills It's not currently on my todo list. Too many other pots on the fire at the moment. One thing we'll have to do is force a response format with the api endpoints, so we should use JSON, and maybe even the JSON-API spec? 

> What are the required Shield setup instructions for API implementation?

That's part of what would need to be determined while we're implementing this. The only thing that are top of mind to me are: 

- ensure either `chain` or `tokens` is the active auth handler, though for our login/register methods, they would not use any authentication.

I would add a configuration option for `apiPath` to the `Auth` config file with a default value of `api/auth` which would then be used in the new `auth()->apiRoutes()` method.

> Is the registration/activation process different from the site/web/traditional flow?

That really depends on the apps setup. As I've said there are two different flows we should account for - both of them already with explanations in the [guides](https://github.com/codeigniter4/shield/tree/develop/docs/guides). I can see a use for register/login/logout API endpoints, though, that would basically be the same logic as the existing elements, but would return JSON. 

> When/where is the appropriate time/place to use auth('tokens') instead of auth() ?

The argument in the the `auth()` helper selects the handler to use (either `session` or `tokens`). So if more than one handler is active, like with the `chain` auth flow, and you need to use a specific handler, then you need to specify which one by passing its alias in. For example, if you have chain auth on because you want to use the same codebase for the admin area and the api, then when you're in the admin you might specify `auth('session')` while working in the api would need to access the Access Tokens directly, so you'd use `auth('tokens')`. 

> What is/are the appropriate way(s) to log a user out of the application completely?

The only way to "log someone out" of using an access token is to [revoke the access token](https://github.com/codeigniter4/shield/blob/develop/docs/authentication.md#revoking-access-tokens)

**Reply by sammyskills** [reply]
Thanks @lonnieezell for the detailed breakdown. 

Please consider this statement:
> It would be difficult to implement while considering all aspects such as that some users may want to utilize the access tokens for third-party API authorization and not authentication. So we should leave it to developers on how they would like to use the access tokens.

Is this the initial idea you had while creating the PAT feature?

**Comment by ghost** [comment]
@lonnieezell 
@sammyskills 

I understand the initial confusion now and think that an authentication API for access tokens may not be worthwhile. It would be difficult to implement while considering all aspects such as that some developers may want to utilize the access tokens for third-party API authorization and not authentication. So we should leave it to developers on how they would like to use the access tokens.

However, I don't want to give up on the idea of an out of box RESTful API for authentication.

Is there a plan for implementing JWT at some point? That would work much better for this use case.

Opinions?

**Reply by sammyskills** [reply]
There is a JWT authentication PR #195 by @kenjis, but it is still a draft.

I don't know if it is ready for use or not.

**Reply by kenjis** [reply]
I believe that #195 is very close to completion.

**Reply by datamweb** [reply]
@kenjis thanks, I will return to it with a few days delay.

**Reply by sammyskills** [reply]
@kenjis, any ETA on the completion?

**Reply by kenjis** [reply]
What is ETA? If more than two people approve, I'm going to merge it.

**Reply by sammyskills** [reply]
> What is ETA? 

Expected Time of Arrival.

>If more than two people approve, I'm going to merge it.

Thank you for your efforts.

**Comment by ghost** [comment]
I am having trouble wrapping my head around the architecture of Shield. Let's anticipate I (and/or others) would like to use CodeIgniter Shield in their back-end that is consumed by a mobile app or other front-end service. For example, mobile app sends email and password credentials to an endpoint - am I forced to utilize the _session authenticator_ to check email/pass credentials (when wanting to use _other authenticators_ such as access tokens or JWT (#195)? For example, in the [docs](https://codeigniter4.github.io/shield/guides/mobile_apps/), the `auth()->attempt()` method is used to create a session just to be returned an access token. To me, it seems like a bit of a conflict unless I am misunderstanding something. I would appreciate any clarification to such use case or if there is any plans to do this type of auth in a cleaner way.

**Comment by lonnieezell** [comment]
No, you're not forced to use Session authenticator, but if you don't you'll need to roll your own validation of the user for login/registration as it currently stands. 

The process is this: 

1. Session-based authentication is used for all email/password type logins. This is the initial protection to prove who they are and that they're allowed to be given another way to authenticate. This also determines the user account to associate the token with.
2. Once you've authenticated with that handler they're handed a token that is a different way to authenticate. Once they have this token, all authentication for future API calls happens through the Access Token layer until the token expires or is revoked, at which point they would have to prove who they are once again, before being given access to the Access Token method of authentication.

I get that it seems strange to use both at first, but they each have their different use cases in what they protect and how they function.
